classlib/src/main/java/org/teavm/classlib/java/lang/TString.java

```diff
@@ -418,20 +428,20 @@ public int lastIndexOf(TString str) {
     }
 
     public TString substring(int beginIndex, int endIndex) {
-        if (beginIndex > endIndex) {
-            throw new TIndexOutOfBoundsException();
+        int length = charactersLength();
+        if (beginIndex < 0 || beginIndex > endIndex || endIndex > length) {
+            throw new TStringIndexOutOfBoundsException();
         }
         if (beginIndex == endIndex) {
             return EMPTY;
         }
 
-        if (PlatformDetector.isJavaScript()) {
-            var nativeSubstring = substringJS(nativeString(), beginIndex, endIndex);
-            return nativeSubstring != nativeString() ? new TString(nativeSubstring) : this;
+        if (beginIndex == 0 && endIndex == length) {
+            return this;
         }
 
-        if (beginIndex == 0 && endIndex == length()) {
-            return this;
+        if (PlatformDetector.isJavaScript()) {
```

Comment:
With this I can't agree. Previous code seems better

Reply:
Why? Strings are equal only in case when beginIndex == 0 and endIndex == length. Removed comparison makes no sense when this values are different.

Reply:
So let JavaScript do this check

Reply:
I'm assuming that comparing two integers is faster in most cases than comparing 2 strings.

Reply:
I assume that native comparison with either have two referentially identical string and returns `true` by the fast path, or otherwise deals with strings of different sizes, so also avoids full char-by-char check. You can also write a benchmark that tests if there's any difference before and after your change.

Reply:
In this case still 2 "integer" comparisons (first references and second length) are performed.
But ok, I will write benchmark.

Reply:
```java
private static final int TEST_COUNT = 1000_000;
    public static void main(String[] args) {
        String[] array = Stream.generate(() -> UUID.randomUUID().toString()).limit(TEST_COUNT).toArray(String[]::new);
        BitSet bs = BitSet.valueOf(LongStream.range(0, TEST_COUNT / Long.BYTES + 1).map(i -> ThreadLocalRandom.current().nextLong()).toArray());
        int counter = 0;
        for (int i = 0; i < array.length - 1; i++) {
            counter += bs.get(i) ? array[i].length() : array[i].length() - 1;
        }
        long time = System.currentTimeMillis();
        for (int j = 0; j < 1000; j++) {
            for (int i = 0; i < array.length - 1; i++) {
                counter += bs.get(i) ? array[i].substring(0, 36) == array[i] ? 1 : 0
                        : array[i].substring(0, 35) == array[i] ? 1 : 0;
            }
        }
        System.out.println("Comparisons " + (System.currentTimeMillis() - time) + " counter " + counter);
    }
```
Luckily for me new implementation is little bit, but faster. Seems that 2 times comparing integers works faster than reference arithmetics/dereferencing. Results:
|`Browser `|`old substring  `|`new substring  `|
|`Chromium`|`14335,14313,14202`|`12098,11971,11802`|
|`Firefox `|`29022,29476,29179`|`26288,27397,26150`|